ocs/ocs_agent.py

```diff
@@ -988,3 +1002,228 @@ def add_message(self, message, timestamp=None):
+        are only activated if the value (or the substituted default
+        value) is not ``None``.
+
+        """
+        self._checked.add(key)
+        value = self._params.get(key, None)
+        is_unset = value is None and \
+            (treat_none_as_missing or key not in self._params)
+        if is_unset:
+            if isinstance(default, ParamError):
+                raise ParamError(f"Param '{key}' is required and must not be None")
+            value = default
+        if value is not None:
+            if cast is not None:
+                try:
+                    value = cast(value)
+                except:
+                    raise ParamError(f"Param '{key}'={value} could not be cast to {cast}.")
+            if type is not None:
+                # Free cast from int to float.
+                if type is float and isinstance(value, int):
+                    value = float(value)
+                if not isinstance(value, type):
+                    raise ParamError(f"Param '{key}'={value} is not of required type ({type})")
+            if choices is not None:
+                if value not in choices:
+                    raise ParamError(f"Param '{key}'={value} is not in allowed set ({choices})")
+        if check is not None:
+            if not check(value):
+                raise ParamError(f"Param '{key}' failed validity check (see docs?).")
+        return value
+
+    def batch(self, instructions, check_for_strays=True):
+        """
+        Supports the @params decorator ... see code.
+        """
+        params = {}
+        for key, kw in instructions:
+            if key == '_':
+                pass
+            elif key == '_no_check_strays':
+                check_for_strays = False
+            else:
+                params[key] = self.get(key, **kw)
+        if check_for_strays:
+            self.check_for_strays()
+        return params
+
+    def check_for_strays(self, ignore=[]):
+        """Raise a ParamError if there were arguments passed in that have not
+        yet been extracted with .get().  Keys passed in ignore (list)
+        will be ignored regardless.
+
+        """
+        weird_args = [k for k in self._params.keys()
+                      if k not in self._checked and k not in ignore]
+        if len(weird_args):
+            raise ParamError(f"params included unexpected values: {weird_args}")
+
+def param(key, **kwargs):
+    """Decorator for Agent operation functions to assist with checking
+    params prior to actually trying to execute the code.  Example::
+
+      class MyAgent:
+        ...
+        @param('voltage', type=float)
+        @param('delay', default=0., type=float)
+        @inlineCallbacks
+        def set_voltage(self, session, params):
+          ...
+
+    Note the ``@param`` decorators should be all together, and
+    outermost (listed first).  ``@inlineCallbacks``, if in use, should
+    be the last decorator.
```

Comment:
Update docstring.  The params stuff has to be outermost because it's not a proper decorator and actually stores private data directly in the function it's decorating.  Other decorators might obscure that data.

When it comes to "other decorators" ... inlineCallbacks also makes life complicated, but I haven't dug to the bottom of that yet.  That decorator takes a a generator (a function with `yield` in it) and converts it to a Deferred callable somehow.